### PR TITLE
Patterns: Tables

### PR DESCRIPTION
**Pull request by talbs**
https://openedx.atlassian.net/browse/UX-2617

This PR contains styling and semantic examples for basic table renderings as a pattern. Because of the restricted (and repetitive) use of table elements - like tr, th, td, etc - that carry semantic meaning, this PR styles those HTML elements rather than requiring classes on each element, and restricts this table styling to true semantic tables.

- - -

##### Reviewers

* [ ] UI/VIZ/FED - @talbs 
* [ ] A11y - @cptvitamin 

- - -

##### Open questions

* [ ] Should the independent table elements be pulled out of the table extend?
* [ ] Should we add independent sort arrow icons (single up and single down that match the double sort arrows) to show the sort direction?

### CODE REVIEW

**Comment by talbs** [comment]
@frrrances, just opening this PR up to keep your good work going. Let me know if you end up not being able to get this to done. I can take it over with some background and carry it forward to Mergetown. :)

**Comment by frrrances** [comment]
@talbs I think this is in a place where you all can review, and once the comments come in, I am happy to help get it over the line or leave it in your hands, whichever you prefer. 

Things I would love feedback from you about: 
* Sass abstraction - Where do you think should I do... more? ... less? ... did I go horribly wrong?
* Examples - too much or too little?
* Doc - should I update the wiki?

Thanks for getting this PR rolling!

File: _posts/patterns/2015-12-11-tables.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+info:           Tables are used for displaying information in tabular format, where is useful to be able to examine and compare data or information in columns and rows. edX offers several table styles to provide ways to display most tabular data, including an example that contains the full set of supported semantic elements.
+
+---
+
+<h3 class="hd-6 example-set-hd">Basic</h3>
+<div class="example-set">
+  <table class="table">
+    <tr>
+      <th>First</th>
```
Comment (cptvitamin) [inline _posts/patterns/2015-12-11-tables.md]: @frrrances we should specify the scope of any ```<th>``` elements (either ```scope=row``` or ```scope=col``` )

File: _posts/patterns/2015-12-11-tables.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    </tbody>
+  </table>
+</div>
+
+
+<h3 class="hd-6 example-set-hd">Sortable</h3>
+<div class="example-set">
+  <table class="table table-sortable">
+    <tr>
+      <th class="sortable sort-current sort-descending" aria-label="Sorted Descending">
```
Comment (cptvitamin) [inline _posts/patterns/2015-12-11-tables.md]: @frrrances ```<th>``` elements don't have an accessible name so the aria-label won't be interpreted here (but it's a good idea!).  Let's try to inform the user some other way, like using offscreen text.  Perhaps we can toggle the offscreen text value to indicate what pressing the ```<button>``` will do e.g. ```Sort<span class=sr>&nbsp;Ascending</span>```  and then toggle that value to Descending on press

cc @talbs

**Comment by talbs** [comment]
@frrrances, after meeting with @cptvitamin, we think its a good idea to start a Styleguilde for Tables with some good best practices, basic set up, and explanation of some of the options you've set up here. I've started a new wiki page (https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Tables). If you're still active on this, would you mind adding some details there? If so, feel free to adjust the structure/sections to fit what you need to explain.

@cptvitamin feel free to also jump in and add some notes on tables and a11y (or I can crib from places where you already have that information).

**Comment by talbs** [comment]
@frrrances, I'm also rebasing this for you against master. I'm going to squash your commits into 1 to make that easier in the future. I'll keep all of your messages/notes though.

**Comment by talbs** [comment]
Okay, the branch has been rebased. @frrrances, you'll need to do a pull on your end. Hopefully I didn't cause you any issues on your local branch.

File: _posts/patterns/2015-12-11-tables.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+categories:     patterns
+tags:
+- atomic
+- pattern
+- buttons
+
+slug:           tables
+
+url_styles:     "components/_tables.scss"
```
Comment (talbs) [inline _posts/patterns/2015-12-11-tables.md]: Looks like you copied from a bad/old example pattern. This changed to be ``"patterns/_[patternname].scss"`` a little bit ago. Sorry for the screw up.

File: _posts/patterns/2015-12-11-tables.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+description:    A collection of buttons available for using within the edX platform.
```
Comment (talbs) [inline _posts/patterns/2015-12-11-tables.md]: NIT: Mind updating the description? You can generally take the ``info`` field and simplify it. I don't think we're using it anywhere (and I should remove it from all patterns), so you could alternatively just remove it from this pattern's metadata.

File: _posts/patterns/2015-12-11-tables.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+info:           Tables are used for displaying information in tabular format, where is useful to be able to examine and compare data or information in columns and rows. edX offers several table styles to provide ways to display most tabular data, including an example that contains the full set of supported semantic elements.
+
+---
+
+<h3 class="hd-6 example-set-hd">Basic</h3>
+<div class="example-set">
```
Comment (talbs) [inline _posts/patterns/2015-12-11-tables.md]: Might be worth adding some descriptions to some of these more nuanced or complex examples (we do that in the grid pattern). Here's the general markup:

```html
<div class="example-set-description copy-meta">
    <p>[example set up, description, or details]</p>
</div>
```

File: _posts/patterns/2015-12-11-tables.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+    </tr>
+    <tr>
+      <td>Treebeard</td>
+      <td>Fangorn</td>
+      <td>Fangorn Forest</td>
+    </tr>
+  </table>
+</div>
+
+<h3 class="hd-6 example-set-hd">Simplified and compact</h3>
```
Comment (talbs) [inline _posts/patterns/2015-12-11-tables.md]: NIT: should we sync up the terms "cozy" and "compact" here?

File: _posts/patterns/2015-12-11-tables.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+      <td>Treebeard</td>
+      <td>Fangorn</td>
+      <td>Fangorn Forest</td>
+    </tr>
+  </table>
+</div>
+
+<h3 class="hd-6 example-set-hd">Full Semantic Set</h3>
+<div class="example-set">
+  <div class="wrapper-table-scrollable">
```
Comment (talbs) [inline _posts/patterns/2015-12-11-tables.md]: NIT: Would be worthwhile noting that this example is scrollable in the description.

File: _posts/patterns/2015-12-11-tables.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    <table class="table">
+      <caption>Some fantastical towns and their key characteristics</caption>
+      <colgroup class="colgroup">
```
Comment (talbs) [inline _posts/patterns/2015-12-11-tables.md]: Okay, I've never seen this before (and its in HTML4's spec). I assume this is how the cols are getting visually highlighted, right? How does this work? :)

File: _posts/patterns/2015-12-11-tables.md [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+          <td>People</td>
+          <td class="align-right">15</td>
+        </tr>
+        <tr>
+          <th>Fandralore</th>
+          <td>Strong metal ores</td>
+          <td>Armorsmithing</td>
+          <td>Sustainable hunting and prosperous trade</td>
+          <td>Elves</td>
+          <td class="align-right">20</td>
```
Comment (talbs) [inline _posts/patterns/2015-12-11-tables.md]: NIT: Good thing to make a note of in the [Styleguide for Tables](https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Tables)

File: _posts/patterns/2015-12-11-tables.md (hunk 8 above)
Comment (talbs) [inline _posts/patterns/2015-12-11-tables.md]: NIT: Good thing to make a note of in the [Styleguide for Tables](https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Tables).

File: _posts/patterns/2015-12-11-tables.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+        <th class="supplemental">Strengths</th>
+        <th>Primary Inhabitants</th>
+        <th>No. Inhabitants</th>
+      </tr>
+    </thead>
+    <tbody>
+      <tr>
+        <th>Pernrith</th>
+        <td>Ancient water source</td>
+        <td class="supplemental">Low economy</td>
```
Comment (talbs) [inline _posts/patterns/2015-12-11-tables.md]: This is a good thing to make a note of in the [Styleguide for Tables](https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Tables).

File: _posts/patterns/2015-12-11-tables.md (hunk 2 above)
Comment (talbs) [inline _posts/patterns/2015-12-11-tables.md]: Nice accounting for states here and prepping visual styles!

File: pattern-library/sass/patterns/_tables.scss [hunk 11]
```diff
@@ -23,3 +24,148 @@ $table-border-color:                        palette(grayscale, x-light) !default
...
+.table-cozy {
+
+  th,
+  td {
+    padding: $table-condensed-cell-padding;
+  }
+}
+
+// extra roomy
+.table-comfortable {
```
Comment (talbs) [inline pattern-library/sass/patterns/_tables.scss]: NIT: we should sync up the term we use for larger spaced tables (``roomy``, ``comfortable``, ``extra``).

File: pattern-library/sass/patterns/_tables.scss [hunk 12]
```diff
@@ -23,3 +24,148 @@ $table-border-color:                        palette(grayscale, x-light) !default
...
+      font-weight: font-weight(semi-bold);
+    }
+  }
+
+  // can be used on cols, rows, or cells
+  .is-highlighted {
+    @extend %highlight;
+  }
+
+  // can be used on rows or cells only
```
Comment (talbs) [inline pattern-library/sass/patterns/_tables.scss]: Is this needed here or do we have these in a more central place? I did a quick spin around and can't seem to find them, but I think I may have added them in [my Lists PR that got out of hand](https://github.com/edx/ux-pattern-library/blob/talbs/elements-lists/pattern-library/sass/global/_helpers.scss#L166). I'll refactor them once these are both merged.

File: pattern-library/sass/patterns/_tables.scss [hunk 13]
```diff
@@ -23,3 +24,148 @@ $table-border-color:                        palette(grayscale, x-light) !default
...
+
+// responsive table enables `supplemental` columns to hide on small screens
+.table-responsive {
+
+  th.supplemental,
+  td.supplemental {
+    display: none;
+  }
+
+  @include susy-breakpoint($bp-screen-md, $susy-default) {
```
Comment (talbs) [inline pattern-library/sass/patterns/_tables.scss]: This makes a lot of sense, but should we worry about hardcoding the specific breakpoint in here (vs. giving folks the option to say when the table's supplemental bits should collapse)?

File: pattern-library/sass/patterns/_tables.scss [hunk 14]
```diff
@@ -23,3 +24,148 @@ $table-border-color:                        palette(grayscale, x-light) !default
 // ----------------------------
 // #GLOBAL
 // ----------------------------
+
+
+%highlight {
+  background-color: palette(warning, x-trans);
+}
+
+%edx-table-base {
```
Comment (talbs) [inline pattern-library/sass/patterns/_tables.scss]: Why the ``edx-`` namespacing here? We don't do it elsewhere and I understand tables being pretty common, but is this the place to break convention. If so, I can retro-fit it into other patterns. Other places seem to namespace pretty heavily (see http://www.lightningdesignsystem.com/components/buttons), so its not a crazy idea, just more weight on remembering the prefix and small amounts of file size when done across places.

File: pattern-library/sass/patterns/_tables.scss [hunk 15]
```diff
@@ -23,3 +24,148 @@ $table-border-color:                        palette(grayscale, x-light) !default
...
+
+  th,
+  td {
+    border: none;
+    border-bottom: 1px solid $table-border-color;
+  }
+}
+
+// extra tight
+.table-cozy {
```
Comment (talbs) [inline pattern-library/sass/patterns/_tables.scss]: NIT: we should sync up the term we use for tighter spaced tables.

File: pattern-library/sass/patterns/_tables.scss [hunk 16]
```diff
@@ -23,3 +24,148 @@ $table-border-color:                        palette(grayscale, x-light) !default
...
+  // sortable header
+  .sortable {
+    position: relative;
+    @include margin-left(spacing-horizontal(x-small));
+
+    // sorting button
+    .action-sort {
+      @extend .btn-secondary;
+      position: absolute;
+      @include right(0);
```
Comment (talbs) [inline pattern-library/sass/patterns/_tables.scss]: Do you need a vertical positioning property here too - ``top`` or ``left``?

File: pattern-library/sass/patterns/_tables.scss [hunk 17]
```diff
@@ -23,3 +24,148 @@ $table-border-color:                        palette(grayscale, x-light) !default
...
+.table-sortable {
+
+  // sortable header
+  .sortable {
+    position: relative;
+    @include margin-left(spacing-horizontal(x-small));
+
+    // sorting button
+    .action-sort {
+      @extend .btn-secondary;
```
Comment (talbs) [inline pattern-library/sass/patterns/_tables.scss]: You'll need to sync this up with the new names I gave buttons in https://github.com/edx/ux-pattern-library/pull/244. Sorry for the mixup.

File: pattern-library/sass/patterns/_tables.scss [hunk 18]
```diff
@@ -23,3 +24,148 @@ $table-border-color:                        palette(grayscale, x-light) !default
 // ----------------------------
```
Comment (talbs) [inline pattern-library/sass/patterns/_tables.scss]: NIT: We usually break up the Sass partials a bit more based on different types/extensions of patterns (see the [Grid partial](https://github.com/edx/ux-pattern-library/blob/master/pattern-library%2Fsass%2Fpatterns%2F_grid.scss) as an example). It makes it easier to abstract and scan. Mind going through your different cases and syncing that up?

**Comment by talbs** [comment]
@frrrances, I just took a spin through. This looks awesome! Things are solid from visual, interaction, and FED PoVs. I left some thoughts, suggestions, and questions inline. Mind addressing those?

If you don't have time, let me know. I can help get this good work to done. Thanks for going the extra mile!

File: _posts/patterns/2015-12-11-tables.md (hunk 4 above)
Comment (frrrances) [inline _posts/patterns/2015-12-11-tables.md]: What, you don't call tables buttons where you're from?

File: _posts/patterns/2015-12-11-tables.md (hunk 3 above)
Comment (frrrances) [inline _posts/patterns/2015-12-11-tables.md]: Oops! Good catch.

File: _posts/patterns/2015-12-11-tables.md (hunk 2 above)
Comment (frrrances) [inline _posts/patterns/2015-12-11-tables.md]: Dang! That would have been too convenient. I've taken your suggestion @cptvitamin and added extra context to the sort button.

File: _posts/patterns/2015-12-11-tables.md (hunk 1 above)
Comment (frrrances) [inline _posts/patterns/2015-12-11-tables.md]: Good catch - I added scope to the ``<th>``s.

File: _posts/patterns/2015-12-11-tables.md (hunk 5 above)
Comment (frrrances) [inline _posts/patterns/2015-12-11-tables.md]: Good idea! My doc skills aren't great but I tried to describe a bit of what's going on with the examples.

File: _posts/patterns/2015-12-11-tables.md (hunk 6 above)
Comment (frrrances) [inline _posts/patterns/2015-12-11-tables.md]: Sure thing. Cozy it is!

File: _posts/patterns/2015-12-11-tables.md (hunk 7 above)
Comment (frrrances) [inline _posts/patterns/2015-12-11-tables.md]: Added the scrollable bit in the description.

File: _posts/patterns/2015-12-11-tables.md (hunk 8 above)
Comment (frrrances) [inline _posts/patterns/2015-12-11-tables.md]: Whoa! Fancy page. I'll get that filled out. :gift:

File: pattern-library/sass/patterns/_tables.scss (hunk 11 above)
Comment (frrrances) [inline pattern-library/sass/patterns/_tables.scss]: Ah, yeah good catch. I synced those up.

File: pattern-library/sass/patterns/_tables.scss (hunk 12 above)
Comment (frrrances) [inline pattern-library/sass/patterns/_tables.scss]: Okay, I'll delete them from here, but assume the classes exist so I can use them in the tables and reference them in the doc.

File: pattern-library/sass/patterns/_tables.scss (hunk 13 above)
Comment (frrrances) [inline pattern-library/sass/patterns/_tables.scss]: Good point. How would we enable that?

File: pattern-library/sass/patterns/_tables.scss (hunk 14 above)
Comment (frrrances) [inline pattern-library/sass/patterns/_tables.scss]: Good call. I started off calling it table so it seemed weird, but i think it's better without the prefix now so I removed it.

File: pattern-library/sass/patterns/_tables.scss (hunk 16 above)
Comment (frrrances) [inline pattern-library/sass/patterns/_tables.scss]: I wanted it to be inline with the text so I didn't mess with the vertical alignment. Think it's needed?

File: pattern-library/sass/patterns/_tables.scss (hunk 17 above)
Comment (frrrances) [inline pattern-library/sass/patterns/_tables.scss]: Okey doke.

File: pattern-library/sass/patterns/_tables.scss (hunk 18 above)
Comment (frrrances) [inline pattern-library/sass/patterns/_tables.scss]: Good idea! I added more sections.

File: pattern-library/sass/patterns/_tables.scss (hunk 12 above)
Comment (talbs) [inline pattern-library/sass/patterns/_tables.scss]: Sounds good!

**Comment by talbs** [comment]
@frrrances, thanks again for the revisions. With your departure from edX, it looks like IT has removed your Github account from edX's Github organization. This is what's causing you to not be able to push changes to this repo directly any longer.

I've got an IT ticket out to make you a contributor (like we do with our open source contributors). This will make sure that your name/account is associated with making this code/tool more awesome (you'll show up in the contributor list and in graphs/history). But, the workflow for providing code will still need to change, here's the skinny on what to do now:

1. You'll want to make a personal fork of the ``edx/ux-pattern-library`` repo here on Github (through the fork button on the repo's main page). This will mean you'll end up with something like ``frrrances/ux-pattern-library`` on your github repo list
2. From the fork, you should have all branches that our remote repo knows about up until the time of the fork. This would include your most recent successful push of this PR's branch.
3. Switch to that branch locally and sync up any missing commits/work (a lot of your PR feedback addressing). I'm not sure if you can cherry-pick locally from your version of the UXPL to your forked local version of the UXPL, but that would be ideal. Less than ideal would mean you having to manually sync up any of those changes.
4. You'll then need to push to the remote version of your forked repo (i.e. ``frrrances/ux-pattern-library``) and create a new PR against this main repo's master branch.
5. We'll then close down this PR started while you were at edX and transfer any next steps (including merging!!!) to that new PR.

Let me know if you need any help with this (as I'll be going through it soon myself).

- - -

FYI, @andy-armstrong and @dsjen

**Comment by frrrances** [comment]
Thanks for the help, @talbs! Good to get things going on this side of the wall and be an open source contributor! Here's the PR: https://github.com/edx/ux-pattern-library/pull/275 Let me know if there's anything else you need. :ghost:

**Comment by talbs** [comment]
Thanks, @frrrances. I'm closing this in favor of your new PR - https://github.com/edx/ux-pattern-library/pull/275.

**Pull request closed without merging**